File: accommodation_service/src/main/java/huy/project/accommodation_service/infrastructure/repository/adapter/LanguageAdapter.java
package huy.project.accommodation_service.infrastructure.repository.adapter;

import huy.project.accommodation_service.core.domain.dto.request.GetLanguageParams;
import huy.project.accommodation_service.core.domain.dto.response.PageInfo;
import huy.project.accommodation_service.core.domain.entity.LanguageEntity;
import huy.project.accommodation_service.core.port.ILanguagePort;
import huy.project.accommodation_service.infrastructure.repository.ILanguageRepository;
import huy.project.accommodation_service.infrastructure.repository.mapper.LanguageMapper;
import huy.project.accommodation_service.infrastructure.repository.model.LanguageModel;
import huy.project.accommodation_service.infrastructure.repository.specification.LanguageSpecification;
import huy.project.accommodation_service.kernel.utils.PageUtils;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.data.util.Pair;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class LanguageAdapter implements ILanguagePort {
    private final ILanguageRepository languageRepository;

    @Override
    public LanguageEntity save(LanguageEntity language) {
        LanguageModel languageModel = LanguageMapper.INSTANCE.toModel(language);
        return LanguageMapper.INSTANCE.toEntity(languageRepository.save(languageModel));
    }

    @Override
    public LanguageEntity getLanguageByName(String name) {
        return languageRepository.findByName(name)
                .map(LanguageMapper.INSTANCE::toEntity).orElse(null);
    }

    @Override
    public LanguageEntity getLanguageByCode(String code) {
        return languageRepository.findByCode(code)
                .map(LanguageMapper.INSTANCE::toEntity).orElse(null);
    }

    @Override
    public Pair<PageInfo, List<LanguageEntity>> getLanguages(GetLanguageParams params) {
        Pageable pageable = PageUtils.getPageable(params);

        var result = languageRepository.findAll(LanguageSpecification.getLanguages(params), pageable);
        return Pair.of(PageUtils.getPageInfo(result), LanguageMapper.INSTANCE.toListEntity(result.getContent()));
    }

    @Override
    public void deleteLanguageById(Long id) {
        languageRepository.deleteById(id);
    }

    @Override
    public LanguageEntity getLanguageById(Long id) {
        return languageRepository.findById(id)
                .map(LanguageMapper.INSTANCE::toEntity).orElse(null);
    }
}

File: location_service/LocationService.Infrastructure/Repository/Adapter/LanguageAdapter.cs
namespace LocationService.Infrastructure.Repository.Adapter
{
    using System.Threading.Tasks;
    using LocationService.Core.Domain.Entity;
    using LocationService.Core.Port;
    using LocationService.Infrastructure.Repository.Mapper;
    using Microsoft.EntityFrameworkCore;

    public class LanguageAdapter : ILanguagePort
    {
        private readonly LocationDbContext _dbContext;

        public LanguageAdapter(LocationDbContext dbContext)
        {
            _dbContext = dbContext;
        }

        public async Task<LanguageEntity> CreateLanguageAsync(LanguageEntity language)
        {
            var model = LanguageMapper.ToModel(language);
            _dbContext.Languages.Add(model);
            await _dbContext.SaveChangesAsync();
            return LanguageMapper.ToEntity(model);
        }

        public async Task<LanguageEntity> GetLanguageByIdAsync(long id)
        {
            var model = await _dbContext.Languages.FindAsync(id);
            return LanguageMapper.ToEntity(model);
        }

        public async Task<LanguageEntity> GetLanguageByCodeAsync(string code)
        {
            var model = await _dbContext.Languages
                .AsNoTracking()
                .FirstOrDefaultAsync(l => l.Code == code);
            return LanguageMapper.ToEntity(model);
        }

        public async Task<LanguageEntity> GetLanguageByNameAsync(string name)
        {
            var model = await _dbContext.Languages
                .AsNoTracking()
                .FirstOrDefaultAsync(l => l.Name == name);
            return LanguageMapper.ToEntity(model);
        }
    }
}